airflow-core/src/airflow/api/common/mark_tasks.py

```diff
@@ -215,17 +215,30 @@ def set_dag_run_state_to_success(
     if not run_id:
         raise ValueError(f"Invalid dag_run_id: {run_id}")
 
-    # Mark all task instances of the dag run to success - except for teardown as they need to complete work.
+    # Mark all task instances of the dag run to success - except for unfinished teardown as they need to complete work.
     normal_tasks = [task for task in dag.tasks if not task.is_teardown]
+    teardown_tasks = [task for task in dag.tasks if task.is_teardown]
+    unfinished_teardown_tis: list[TaskInstance] = session.scalars(
+        select(TaskInstance).where(
+            TaskInstance.dag_id == dag.dag_id,
+            TaskInstance.run_id == run_id,
+            TaskInstance.task_id.in_([task.task_id for task in teardown_tasks]),
+            or_(TaskInstance.state.is_(None), TaskInstance.state.in_(State.unfinished)),
+        )
+    ).all()
```

Comment:
We only really needs the task_id here, so let’s just select that field instead of the entire TaskInstance object. Also putting the task_ids in a set (instead of list) would be a good idea.